src/Sarif/Visitors/SortingVisitor.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    public class SortingVisitor : SarifRewritingVisitor
+    {
+        private readonly IDictionary<int, int> ruleIndexMap;
+        private readonly IDictionary<int, int> artifactIndexMap;
+
+        public SortingVisitor()
+        {
+            ruleIndexMap = new ConcurrentDictionary<int, int>();
+            artifactIndexMap = new ConcurrentDictionary<int, int>();
+        }
+
+        public override SarifLog VisitSarifLog(SarifLog node)
+        {
+            SarifLog current = base.VisitSarifLog(node);
+            if (node?.Runs != null)
+            {
+                current.Runs = current.Runs.OrderBy(r => r, RunSortingComparer.Instance).ToList();
+            }
+            return current;
+        }
+
+        public override Run VisitRun(Run node)
+        {
+            if (node?.Artifacts != null)
+            {
+                IDictionary<Artifact, int> oldIndexes = new Dictionary<Artifact, int>(capacity: node.Artifacts.Count);
+                // save old indexes
+                for (int i = 0; i < node.Artifacts.Count; i++)
+                {
+                    if (!oldIndexes.ContainsKey(node.Artifacts[i]))
```

Comment:
>ContainsKey [](http://example.com/codeflow?start=36&length=11)

either use the indexer here or just use TryUpdate on ConcurrentDictionary

Reply:
oldIndexes[node.Artifacts[1]]

Reply:
this dictionary is a local variable, I think it's thread-safe and does not need ConcurrentDictionary.

use indexer to access dictionary will need to handle `KeyNotFoundException` so I use `ContainsKey` or `TryGetValue` to check if key exists.

in latest change, move this to common function. please review and let me know what do you think.